WebCryptoAPI/wrapKey_unwrapKey/wrapKey_unwrapKey.https.any.js

```diff
@@ -127,112 +129,101 @@
             formats = ["raw", "jwk"]
         }
 
-        return Promise.all(formats.map(function(fmt) {
-            var originalExport;
-            return subtle.exportKey(fmt, toWrap.key).then(function(exportedKey) {
-                originalExport = exportedKey;
-                const isPossible = wrappingIsPossible(originalExport, wrapper.parameters.name);
-                promise_test(function(test) {
-                    if (!isPossible) {
-                        return Promise.resolve().then(() => {
-                            assert_false(false, "Wrapping is not possible");
-                        })
+        formats.forEach(async(fmt) => {
+            var originalExport = await subtle.exportKey(fmt, toWrap.key);
+            if (!wrappingIsPossible(originalExport, wrapper.parameters.name))
+                return;
```

Comment:
```suggestion
        await Promise.all(formats.map(async(fmt) => {
            var originalExport = await subtle.exportKey(fmt, toWrap.key);
            if (!wrappingIsPossible(originalExport, wrapper.parameters.name))
                return;
```

Reply:
Would this change ensure that the subtests are going to be executed always in the same order ? That's one of the issues I wanted to solve with this refactoring.

Reply:
Ah, no. For that, you'd have to either make this a sequential loop (`for (const fmt of formats) ...`) or move the `await subtle.exportKey` to the setup (e.g. inside `generateKeysToWrap()`?) and reference the value here (the latter is probably better, in order to not slow down the test suite).

If you do the latter and make `testWrapping` not use any `await`s, then you also don't have to make the changes I suggested, so then you probably have a cleaner diff.

Reply:
Actually, I think it's not relevant if the setup phase is synchronous or asynchronous; it's the execution of the testWrapping what really matters. We would just need to wain until all the promises in the setup phase are solved.

Reply:
Yes, that's why I suggested to move this `await subtle.exportKey` from `testWrapping` to the setup phase :)

Reply:
I have moved the exportKey call to the generateKeysToWrap, as suggested. 

However, the test is still not solid enough. I believe it's too complex, especially regarding the generation of the test cases. The number of test cases varies between executions; it's specially concerning

Reply:
> However, the test is still not solid enough. I believe it's too complex, especially regarding the generation of the test cases. The number of test cases varies between executions; it's specially concerning

https://github.com/web-platform-tests/wpt/pull/38247#issuecomment-1419329361

Reply:
Ah, I see! Thanks for remembering that :sweat_smile:

Then, actually, probably the best thing to do would be to either:
1. Move that particular check inside the test, despite what I said before, and just pass the test without doing anything if the key isn't guaranteed to be able to be wrapped.
2. Retry key generation in this specific case until we generate one that can be wrapped? (Probably not ideal)
3. Propose a spec change to mandate the padding implemented by Node.js, so that the key can always be wrapped :smile:

Reply:
I have tested this PR in a variety of bot in WebKit and there is only problems with RSA related tests. We are speculating with the non-determinisc key generation as the root cause of this inconsistencies on the exportedKey's length.

Reply:
> 
> https://w3c.github.io/webcrypto/#:~:text=For%20key%20formats%20that%20offer%20flexibility%20in%20serialization%20of%20a%20given%20key%20(for%20example%20JWK)%2C%20implementations%20may%20choose%20to%20adapt%20the%20serialization%20to%20the%20constraints%20of%20the%20wrapping%20algorithm.
> 

The test case that is passing the isWrappingPossible occasionally is RSA-PSS and pkcs8 export.

Reply:
> Ah, I see! Thanks for remembering that 😅
> 
> Then, actually, probably the best thing to do would be to either:
> 
>     1. Move that particular check inside the test, despite what I said before, and just pass the test without doing anything if the key isn't guaranteed to be able to be wrapped.
> 
>     2. Retry key generation in this specific case until we generate one that can be wrapped? (Probably not ideal)
> 
>     3. Propose a spec change to mandate the padding implemented by Node.js, so that the key can always be wrapped 😄

4. Use import rather than generateKey, as we do in many other tests. I know it makes the tests less effective because we are using always the same keys, but this are just functional tests, aren't they ? 

If we want to define kind of fuzzy tests, we may need to define them in a different way.

Reply:
Right, that also works 👍

However.. I don't think the length of generated RSA keys is supposed to vary, unlike P-521 keys, for example. So it may then be an actual issue in the implementation, if e.g. it's sometimes generating a key with fewer bits than it's supposed to?
(I can double check tomorrow, on mobile rn.)

Reply:
> I don't think the length of generated RSA keys is supposed to vary

Both keys are modulusLength 1024. Different length after JSON.stringify 848 vs 846, sometimes 847

```
848 {
  kty: 'RSA',
  n: 'sG8i7gYjubd5o0sdBt-_ehrZh4SpQM_8pM6G9AiMUOcP1GqTi2FpfYS8Ch5ETt5ukluwU4twPJWGAJLBuJnW6MnkUzNw0NGTheC3EvB7mmuEjz-zOfRnzMSi7MyPKBSN8gnJ8AoDkJ_aEQSkXFABMkoX_bl8T160XOhhjLoNqtk',
  e: 'AQAB',
  d: 'h-jUi26HTqSbOsK91XG-Wh0cvFvg1WBiScIeCxYaoKeddg-zIIclb98zyseN7wNBQdUHexPKnaDA3Ogv-jfnZ5t98k6CF5esiieDIWHgEOLCJOBMn49CfBtH5xHvsJ7Fe79jVF25XPy6ueZB0Yn0T6i03_BGxuj2mVsyJi8h5pE',
  p: '3i4Rg6b11LG9FJbJRAg5Aiedt6eJnsiijHcK51c6qDOL1toi0Z9wdSgdNcLJsJm9cu4JkIHfLdVhZ_09k_yW1w',
  q: 'y0pwmbfFo_UvwcZi1y5XW1obCihe4DrYXa0EOe3mQCom46cQ0gLAF8Br7EYJ6Xh8iwlf-219wxFlVSEaON6Fzw',
  dp: 'sPqpYOFh-OcL4paqA9M49gpwi5ac_Sp8sL8ypEzLoKhOv_pKugfzIJ-vhN-AoXgVfLM4Uy_ggzXxvU_v-S2_SQ',
  dq: 'GHvFpe3JIt8Dum2_N7VAFiN7ADUugAj2SAnvaV6qjW3GM0LKttBX_ERTeyhQR0sYjNI7OhfMvrKYprEO7Gx4Zw',
  qi: 'CnatzTFW2dY8m-ddOjkqzFYj-XEO5v81T52v19U_MWJYGAnyY9NqOfF3g711j459xXxz3EQIy3951Xr7amnv9A'
}
846 {
  kty: 'RSA',
  n: 'v6wvJHdRelrefDLOf9Ey9dYtD1I5PEp3mPElmuT-tmMGfnem0eHL7wn_xNg-bbvrMkryYzliGxPhZv2N5w15Ro_BPM6BHxSWVPVnFwvRgbjOjOyiHjOslN_PYjirAhrOV6k7zxJ8Un4qSkUaO9lbO43JoC6Af27vcnf1_D4Mlk0',
  e: 'AQAB',
  d: 'i8cUhgJhr9nuoCIpB3tXL4JmT-MFjFs-BpHte5z9x6JFjRrS1AJXR1ZCaJnORRYiugV_6C5jxh1kZkgIIPRpeHDW3Af8ZMaJcLNdS1dRItn4t5QV7LYdQJGVaRKxhLtsqsKTnOnzJjHmdyursWKP_bhIgfVtql1pcBeF22D7GIE',
  p: '7hRSRa20xCo7XyuzbPqPi7dBoJAoGSn3PvZ63G_7mr7EQG51Ncn_3LukF1Qp8x8UXK7zDoVuImkxtlpgcf4bbQ',
  q: 'zhmgKqWSI7zc68IYI8MnMCxaSRbbs5TJVRNGTs89sBhevl6B-pBPA8XV3cwrrFMWe88k129qb1ZBsaE_2tO6YQ',
  dp: 'a-DJZslPH5QDUxx6MOdYI_blxduMT9q0LdOB2GDloTjPtGJYugbN6TcdI3T_jSqID3c2q13Kic0guYA6M2zx',
  dq: 'h4Mb586Tx0G_9JfbLXB8jNUszA_mhy5WLsUMnnbvNaDgcsVLK0fFXFKxKzO6p4S-LMvnimtsjSYm7bd7d9NtIQ',
  qi: 'L6wvGs_0K0mjFdz6JmWYDm_qsAkbJEmQdZdaDzRK_8MAwkI_4INUAKKAoCB-Yb4v1U7RDO17An7et_TpSCFsJg'
}
```

Same thing happens with 4096 bit length, and with 2048 i only get non-multiple of 8 JSON.stringified lengths

Reply:
Right, I can reproduce also with PKCS8 that the length varies. The public key is always 4096 bits but the private params may vary in length somewhat, I guess that's fine indeed.

And, unfortunately in PKCS8 it's not possible to add padding, so option 3 is probably out unless we do something like https://www.rfc-editor.org/rfc/rfc5649.html :)

Importing rather than generating a test case (perhaps one which _can_ be wrapped?) would be easiest indeed.

Reply:
I'm trying the "importing" approach right  now. Lets see if I manage to produce a meaningful test.

Reply:
Thanks! 🙏